policy/parser.go

```diff
@@ -0,0 +1,615 @@
+	if err != nil {
+		return err
+	}
+	if docNode.Kind != yaml.DocumentNode {
+		return fmt.Errorf("got yaml node of kind %v, wanted mapping node", docNode.Kind)
+	}
+	return nil
+}
+
+func newParserImpl(visitor TagVisitor, info *ast.SourceInfo, src *Source, iss *cel.Issues) *parserImpl {
+	return &parserImpl{
+		visitor: visitor,
+		info:    info,
+		src:     src,
+		iss:     iss,
+	}
+}
+
+type parserImpl struct {
+	id      int64
+	visitor TagVisitor
+	info    *ast.SourceInfo
+	src     *Source
+	iss     *cel.Issues
+}
+
+func (p *parserImpl) NextID() int64 {
+	p.id++
+	return p.id
+}
+
+func (p *parserImpl) NewPolicy(node *yaml.Node) (*Policy, int64) {
+	policy := &Policy{}
+	policy.source = p.src
+	policy.info = p.info
+	policy.semantic = firstMatch
+	id := p.CollectMetadata(node)
+	return policy, id
+}
+
+func (p *parserImpl) NewRule(node *yaml.Node) (*Rule, int64) {
+	r := &Rule{}
+	id := p.CollectMetadata(node)
+	return r, id
+}
+
+func (p *parserImpl) NewVariable(node *yaml.Node) (*Variable, int64) {
+	v := &Variable{}
+	id := p.CollectMetadata(node)
+	return v, id
+}
+
+func (p *parserImpl) NewMatch(node *yaml.Node) (*Match, int64) {
+	m := &Match{}
+	id := p.CollectMetadata(node)
+	return m, id
+}
+
+func (p *parserImpl) NewString(node *yaml.Node) ValueString {
+	id := p.CollectMetadata(node)
+	nodeType := p.assertYamlType(id, node, yamlString, yamlText)
+	if nodeType == nil {
+		return ValueString{ID: id, Value: "*error*"}
+	}
+	if *nodeType == yamlText {
+		return ValueString{ID: id, Value: node.Value}
+	}
+	if node.Style == yaml.FoldedStyle || node.Style == yaml.LiteralStyle {
+		col := node.Column
+		line := node.Line
+		txt, found := p.src.Snippet(line)
+		indent := ""
+		for len(indent) < col-1 {
+			indent += " "
+		}
+		var raw strings.Builder
+		for found && strings.HasPrefix(txt, indent) {
+			line++
+			raw.WriteString(txt)
+			txt, found = p.src.Snippet(line)
+			if found && strings.HasPrefix(txt, indent) {
+				raw.WriteString("\n")
+			}
+		}
+		offset := p.info.OffsetRanges()[p.id]
+		offsetStart := offset.Start - (int32(node.Column) - 1)
+		p.info.SetOffsetRange(p.id, ast.OffsetRange{Start: offsetStart, Stop: offsetStart})
+		return ValueString{ID: id, Value: raw.String()}
+	}
+	return ValueString{ID: id, Value: node.Value}
+}
+
+func (p *parserImpl) CollectMetadata(node *yaml.Node) int64 {
+	id := p.NextID()
+	line := node.Line
+	col := int32(node.Column)
+	switch node.Style {
+	case yaml.DoubleQuotedStyle, yaml.SingleQuotedStyle:
+		col++
+	}
+	offsetStart := int32(0)
+	if line > 1 {
+		offsetStart = p.info.LineOffsets()[line-2]
+	}
+	p.info.SetOffsetRange(id, ast.OffsetRange{Start: offsetStart + col - 1, Stop: offsetStart + col - 1})
+	return id
+}
+
+func (p *parserImpl) parsePolicy(ctx ParserContext, node *yaml.Node) *Policy {
+	ctx.CollectMetadata(node)
+	policy, id := ctx.NewPolicy(node)
+	if p.assertYamlType(id, node, yamlMap) == nil {
+		return policy
+	}
+	for i := 0; i < len(node.Content); i += 2 {
```

Comment:
I'm reasonably certain that the parser would enforce the yaml tags to be a key-value pair, but it would help to either add a check for this or at least a comment on why `+=2` is safe.